### ISSUE TITLE
Agent stats logging

### ISSUE THREAD

**Issue by theSoenke** [rlgraph/rlgraph#64]
There currently seems to be no support for logging the progress of the agent when using the `ApexExecutor`. For example the rewards after each episode. For the `SingleThreadedWorker` it is possible to pass a callback that is executed after each episode. This does not seems to be implemented for the `ApexExecutor` as far as I can see it.

As a workaround I now aggregate and log the results in the report interval: https://github.com/rlgraph/rlgraph/blob/master/rlgraph/execution/ray/ray_executor.py#L192

Would be great to have better logging support in the library. Adding these stats to Tensorboard would also be nice

**Comment by michaelschaarschmidt** [comment]
Hey,

thanks for the issue. That's a reasonable point, will see if the API can be improved to return a stats dict per step call, and the have the workers not return just step count but a stat dict. 

Summaries should work in the learner similar to:

https://github.com/rlgraph/rlgraph/blob/master/examples/dqn_cartpole_with_tf_summaries.py

does it not?

**Comment by theSoenke** [comment]
> will see if the API can be improved to return a stats dict per step call, and the have the workers not return just step count but a stat dict.

That sounds good and probably will be faster than fetching the results from all workers each interval.

> Summaries should work in the learner similar to: https://github.com/rlgraph/rlgraph/blob/master/examples/dqn_cartpole_with_tf_summaries.py
does it not?

I've tested the summaries with the `ApexExectutor` and I saw no results in Tensorboard

**Comment by michaelschaarschmidt** [comment]
I think the reason summaries do not work is because 
```
          "disable_monitoring": true,
```
is on in most configs. That turns off all profiling in the learner for performance reasons (really slows down learning typically.

I will shortly merge a branch which always los min, mean and max reward across all finished episodes within the recent iteration. That should help keep track of progress.

**Comment by theSoenke** [comment]
`"disable_monitoring": true` really was the issue. Logging the rewards also works for me. Thanks for implementing it!

**Comment by theSoenke** [comment]
Might have been a little bit too fast with closing this. I've set `"disable_monitoring": false,` for the execution and worker execution spec and Tensorboard only shows the graph but nothing else

```json
{
    "type": "apex",
    "discount": 0.99,
    "n_step": 3,
    "memory_spec": {
        "type": "prioritized_replay",
        "capacity": 3000
    },
    "saver_spec": {
        "save_secs": 600,
        "max_checkpoints": 5
    },
    "preprocessing_spec": [
        {
            "type": "image_resize",
            "width": 84,
            "height": 84,
            "scope": "image_resize"
        },
        {
            "type": "convert_type",
            "to_dtype": "float",
            "scope": "convert_type"
        },
        {
            "type": "divide",
            "divisor": 255.0,
            "scope": "divide"
        },
        {
            "type": "sequence",
            "sequence_length": 4,
            "batch_size": 1,
            "add_rank": false,
            "scope": "sequence"
        }
    ],
    "network_spec": [
        {
            "type": "conv2d",
            "filters": 16,
            "kernel_size": 8,
            "strides": 4,
            "padding": "same",
            "activation": "relu",
            "scope": "conv1"
        },
        {
            "type": "conv2d",
            "filters": 32,
            "kernel_size": 4,
            "strides": 2,
            "padding": "same",
            "activation": "relu",
            "scope": "conv2"
        },
        {
            "type": "conv2d",
            "filters": 256,
            "kernel_size": 11,
            "strides": 1,
            "padding": "valid",
            "activation": "relu",
            "scope": "conv3"
        },
        {
            "type": "reshape",
            "flatten": true
        }
    ],
    "policy_spec": {
        "type": "dueling-policy",
        "units_state_value_stream": 512,
        "action_adapter_spec": {
            "pre_network_spec": [
                {
                    "type": "dense",
                    "units": 512
                }
            ]
        }
    },
    "exploration_spec": {
        "epsilon_spec": {
            "decay_spec": {
                "type": "constant_decay",
                "constant_value": 0.0
            }
        }
    },
    "execution_spec": {
        "gpu_spec": {
            "gpus_enabled": true,
            "max_usable_gpus": 1,
            "allow_memory_growth": true
        },
        "disable_monitoring": false,
        "session_config": {
            "log_device_placement": false,
            "allow_soft_placement": true,
            "device_count": {
                "CPU": 1
            },
            "inter_op_parallelism_threads": 1,
            "intra_op_parallelism_threads": 1
        },
        "ray_spec": {
            "executor_spec": {
                "redis_address": "127.0.0.1:6379",
                "num_cpus": null,
                "num_gpus": null,
                "weight_sync_steps": 400,
                "replay_sampling_task_depth": 4,
                "env_interaction_task_depth": 2,
                "num_worker_samples": 198,
                "learn_queue_size": 16,
                "num_sample_workers": 8,
                "num_replay_workers": 1,
                "num_cpus_per_replay_actor": 1
            },
            "worker_spec": {
                "execution_spec": {
                    "gpu_spec": {
                        "gpus_enabled": false
                    },
                    "disable_monitoring": false,
                    "session_config": {
                        "allow_soft_placement": true,
                        "device_count": {
                            "CPU": 1
                        },
                        "inter_op_parallelism_threads": 1,
                        "intra_op_parallelism_threads": 1
                    }
                },
                "num_worker_environments": 6,
                "num_background_envs": 0,
                "worker_frame_skip": 1,
                "env_internal_frame_skip": 4,
                "n_step_adjustment": 3,
                "worker_executes_postprocessing": true,
                "sample_exploration": false,
                "ray_constant_exploration": true,
                "num_cpus": 1,
                "worker_executes_exploration": true
            },
            "apex_replay_spec": {
                "memory_spec": {
                    "capacity": 1000000,
                    "alpha": 0.6,
                    "beta": 0.4
                },
                "clip_rewards": false,
                "min_sample_memory_size": 50000
            }
        }
    },
    "observe_spec": {
        "buffer_size": 3000
    },
    "update_spec": {
        "do_updates": true,
        "update_interval": 10000,
        "steps_before_update": 50000,
        "batch_size": 512,
        "sync_interval": 500000
    },
    "optimizer_spec": {
        "type": "adam",
        "learning_rate": 0.0001,
        "clip_grad_norm": 40
    }
}

```

**Comment by michaelschaarschmidt** [comment]
cc'ing @sven1977 who looked at summaries recently

**Comment by sven1977** [comment]
Hi, I see that there is currently no specification of any `summary_spec` in your Agent. This parameter that gets passed to Agent directly can have the following keys:

`summary_regexp`: with which you can specify, which variables should get summaries attached (by Perl regexp matching the global-scope component/variable name).

`directory`: The directory, in which summaries are written.

Summaries so far only work for variables, not for specific ops inside Components' graph_fn. Janislav is currently looking at this and we'll have a generic solution soon that allows you to summarize any op you would like.
